circom-proving/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+use ark_bn254::{Bn254, Fr};
+use ark_crypto_primitives::Error;
+use thiserror::Error;
+use ark_groth16::{
+	verify_proof as ark_verify_proof,
+	PreparedVerifyingKey,
+	Proof as ArkProof,
+	VerifyingKey as ArkVerifyingKey,
+	create_proof_with_reduction_and_matrices,
+	prepare_verifying_key,
+	ProvingKey,
+	VerifyingKey
+};
+use ark_circom::{CircomReduction, WitnessCalculator};
+use std::{
+	sync::Mutex,
+	// convert::{TryFrom, TryInto},
+};
+use ark_std::{rand::thread_rng, UniformRand};
+use num_bigint::{BigInt};
+use cfg_if::cfg_if;
+use ark_relations::r1cs::{ConstraintMatrices, SynthesisError};
+use ark_serialize::CanonicalDeserialize;
+use wasmer::{Module, Store};
+use once_cell::sync::OnceCell;
+// use ark_std::vec::Vec;
+use arkworks_native_gadgets::to_field_elements;
+
+#[derive(Error, Debug)]
+pub enum ProofError {
+	#[error("Error reading circuit key: {0}")]
+	CircuitKeyError(#[from] std::io::Error),
+	#[error("Error producing witness: {0}")]
+	WitnessError(color_eyre::Report),
+	#[error("Error producing proof: {0}")]
+	SynthesisError(#[from] SynthesisError),
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+static WITNESS_CALCULATOR: OnceCell<Mutex<WitnessCalculator>> = OnceCell::new();
+// Initializes the witness calculator using a bytes vector
+#[cfg(not(target_arch = "wasm32"))]
+pub fn circom_from_raw(wasm_buffer: Vec<u8>) -> &'static Mutex<WitnessCalculator> {
+	WITNESS_CALCULATOR.get_or_init(|| {
+		let store = Store::default();
+		let module = Module::new(&store, wasm_buffer).unwrap();
+		let result =
+			WitnessCalculator::from_module(module).expect("Failed to create witness calculator");
+		Mutex::new(result)
+	})
+}
+
+// Initializes the witness calculator
+#[cfg(not(target_arch = "wasm32"))]
+pub fn circom_from_folder(wasm_path: &str) -> &'static Mutex<WitnessCalculator> {
+	// We read the wasm file
+	let wasm_buffer = std::fs::read(wasm_path).unwrap();
+	circom_from_raw(wasm_buffer)
+}
+
+pub fn generate_proof<const N: usize>(
```

Comment:
Can you break proving out into a separate file and put a feature flag on the `mod`? Best to keep code that can't go in the runtime separate.

Reply:
This entire folder `circom-proving` is already meant to not be used in the runtime. It is only meant for testing. Technically speaking, we can get rid of the `verify_proof` function in this file, since it is identical to that in `primitives/verifying/circom.rs`.